### PR TITLE
Add constraints to FixedWidthInteger.Stride and .Magnitude

### PR DESCRIPTION
**Pull request by stephentyrone**
Add the constraint that these associatedtypes themselves conform to FixedWidthInteger and to SignedInteger and UnsignedInteger, respectively. These are effectively part of the semantic requirement of the protocol already, because the requirements on .min and .max effectively force FixedWidthInteger to be twos-complement, which requires Magnitude not be Self for signed types. Also, in practice it's generally necessary to have these two constraints in order to effectively use the FixedWidthInteger protocol anyway; witness that by adding them to the protocol, we eliminate them as constraints from a number of extensions.

Resolves [SR-8156](https://bugs.swift.org/browse/SR-8156).

### CODE REVIEW

**Comment by stephentyrone** [comment]
@swift-ci test compiler performance

**Comment by stephentyrone** [comment]
Also, @airspeedswift if there's no perf regression, are you ok with pulling this into 4.2?

**Comment by stephentyrone** [comment]
@swift-ci please test

File: stdlib/public/core/Integers.swift.gyb [hunk 1]
```diff
@@ -2271,7 +2271,8 @@ extension BinaryInteger {
 /// other arithmetic methods and operators.
 public protocol FixedWidthInteger :
   BinaryInteger, LosslessStringConvertible
-  where Magnitude : FixedWidthInteger
+  where Magnitude : FixedWidthInteger & UnsignedInteger,
+           Stride : FixedWidthInteger & SignedInteger
```
Comment (moiseev) [inline stdlib/public/core/Integers.swift.gyb]: nitpick: we usually don't align the `:`s.

Comment (stephentyrone) [reply stdlib/public/core/Integers.swift.gyb]: boring =) I'll fix before committing.

**Comment by moiseev** [comment]
:shipit:

**Comment by airspeedswift** [comment]
How plausible is it that this could break source? Are there possible ways that you could be relying on not fulfilling these constraints and still get working code?

**Comment by airspeedswift** [comment]
@swift-ci please test source compatibility

**Comment by airspeedswift** [comment]
Linux failure unrelated?

@swift-ci please test linux platform

**Comment by stephentyrone** [comment]
@airspeedswift Re: breaking source. It's *relatively* low-risk because practically any concrete type conforming to `FixedWidthInteger` will already have the conformances. You would have to have defined a concrete fixed-size integer type whose stride was *not* a signed fixed-size integer, which is really going out of your way, or whose magnitude was not UnsignedInteger--this second possibility is the only one that worries me a little bit, but it's still pretty niche. The main risk is that if someone implemented their own fixed-size bignum using a sign-magnitude representation, they might have defined `Magnitude == Self`. This is probably a violation of the semantics documented for `FixedWidthInteger`, but in a subtle way, and it will have "worked" until now.

For anyone not defining their own integer types, it's a non-issue.

**Comment by xwu** [comment]
@stephentyrone Would it help to make explicit in the documentation that `FixedWidthInteger` types are two's complement? I think a bunch of generic code makes that assumption already, and it would make the `.Magnitude` constraint a no-brainer.

**Comment by stephentyrone** [comment]
@swift-ci Smoke test and merge

**Comment by stephentyrone** [comment]
@xwu Yeah, I'll give some thought to what the right way to document that is. We don't strictly require that the actual internal representation be twos-complement, rather that the API surface behaves as though it were.

**Comment by stephentyrone** [comment]
@swift-ci smoke test

**Comment by stephentyrone** [comment]
@xwu further improvement to documentation of `FixedWidthInteger`: https://bugs.swift.org/browse/SR-8189

**Comment by dabrahams** [comment]
LGTM basically.  The one question I have is whether it's right to constrain `Stride` to being signed?

**Comment by stephentyrone** [comment]
@dabrahams All strides are signed: `associatedtype Stride : Comparable, SignedNumeric`. The new requirement is that the stride of a fixed-width integer is a fixed-width integer.

File: stdlib/public/core/Integers.swift.gyb [hunk 2]
```diff
@@ -2271,7 +2271,8 @@ extension BinaryInteger {
 /// other arithmetic methods and operators.
 public protocol FixedWidthInteger :
   BinaryInteger, LosslessStringConvertible
-  where Magnitude : FixedWidthInteger
+  where Magnitude : FixedWidthInteger & UnsignedInteger,
+        Stride : FixedWidthInteger & SignedInteger
```
Comment (dabrahams) [inline stdlib/public/core/Integers.swift.gyb]: @stephentyrone I'm referring to the line above.  Either "`& SignedInteger`" is redundant—in which case you should drop it—or you're constraining `Stride` to be signed here.  Or, I'm missing something ;-)

Comment (stephentyrone) [reply stdlib/public/core/Integers.swift.gyb]: What we get from `Strideable` is `SignedNumeric`. `SignedInteger` is a *very* tiny refinement of `BinaryInteger, SignedNumeric`, but I think it's the correct requirement semantically.

Comment (dabrahams) [reply stdlib/public/core/Integers.swift.gyb]: Okay then!

**Pull request merged**